ui: bug fixes

maxlang opened the pull request:
<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/7783)

<!-- Reviewable:end -->

Commit:
Expand db links to fit the table cell.

Fixes #7054

Commit:
Add warning for low number of nodes. Includes tooltip warning design.

Fixes #5978

Commit:
Fix nvd3 tooltip precision.

Fixes #7592

tamird commented:
Please include "fixes #xxxx" in each commit message rather than the PR description.

Can you include some screenshots where applicable?

---

Reviewed 4 of 4 files at r1, 6 of 6 files at r2, 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[ui/app/components/sortabletable.tsx, line 124 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783#-KMV3ros7nWvqpUETLsB:-KMV3ros7nWvqpUETLsC:-704299366) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f5baae360860332b4a765e6890bb650bb9a5f14/ui/app/components/sortabletable.tsx#L124)):*
> ```XML
>           return <tr key={rowIndex}>
>             {_.map(columns, (c: SortableColumn, colIndex: number) => {
>               let content = c.cell(rowIndex);
> ```

how come this had to change?

---

*[ui/app/components/stackedgraph.tsx, line 79 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783#-KMV4-vvjSgLtEAVq5UZ:-KMV4-vwxMGfKXOw-rUC:604978585) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec759c23faa2d0199f25ea7a1fb0db08011d2db/ui/app/components/stackedgraph.tsx#L79)):*
> ```XML
>       .margin(CHART_MARGINS);
> 
>     (this.chart as any).showControls(false);
> ```

revert this.

---

*[ui/app/components/visualization.tsx, line 9 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783#-KMV4787pd3LSgpq3lHW:-KMV4787pd3LSgpq3lHX:-940453248) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec759c23faa2d0199f25ea7a1fb0db08011d2db/ui/app/components/visualization.tsx#L9)):*
> ```XML
>   subtitle?: string;
>   tooltip?: string;
>   // If warning or warningTitle exist, they are appended to the tooltip and the icon is changed to the warning icon.
> ```

wrap comments at 80 please

---

*[ui/app/containers/clusterOverview.tsx, line 55 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783#-KMV4LYVUGhvMH3l1SGW:-KMV4LYVUGhvMH3l1SGX:759960366) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec759c23faa2d0199f25ea7a1fb0db08011d2db/ui/app/containers/clusterOverview.tsx#L55)):*
> ```XML
>           <Visualization title={ (totalNodes === 1) ? "Node" : "Nodes" }
>                          tooltip="The total number of nodes in the cluster."
>                          warning={ totalNodes < 3 ? `A ${totalNodes} node cluster cannot survive node failure.` : "" }
> ```

s/survive/tolerate/

this also seems like an odd place to put this warning - if any range has fewer than 3 nodes it should trigger a warning (this can happen via bad zone config, for instance). perhaps we could move this to a per-range warning?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783#-:-KMV3dY3NdiQmbfxaILS:-961915312)*
<!-- Sent from Reviewable.io -->

maxlang commented:
Whoops, forgot to include this:

Warning for low number of nodes.
![screen shot 2016-07-12 at 11 53 57 am](https://cloud.githubusercontent.com/assets/3691886/16777745/c7efa3dc-4838-11e6-9a0a-443a8350dc42.png)

maxlang commented:
Added issue #'s

cc @petermattis re low # of nodes warning

---

Review status: 2 of 10 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[ui/app/components/sortabletable.tsx, line 124 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783#-KMV3ros7nWvqpUETLsB:-KMVqBybAHshles-A2lH:1439802904) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f5baae360860332b4a765e6890bb650bb9a5f14/ui/app/components/sortabletable.tsx#L124)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> how come this had to change?
</details>
In order for the links to fill the cell, I had to move some things from the `td` table cell element to the child element. If the child element isn't a "react element" that means it isn't wrapped, so there's nothing for the new styles to affect and the table cells padding is off.

This reminded me to wrap everything except links, since other react elements could be converted to text or numbers as well.

---

*[ui/app/components/visualization.tsx, line 9 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783#-KMV4787pd3LSgpq3lHW:-KMVpkkJFQX9viJwAP16:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec759c23faa2d0199f25ea7a1fb0db08011d2db/ui/app/components/visualization.tsx#L9)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> wrap comments at 80 please
</details>
Done.

---

*[ui/app/containers/clusterOverview.tsx, line 55 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783#-KMV4LYVUGhvMH3l1SGW:-KMVpuc7TYYye92uSCAw:-1605311027) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec759c23faa2d0199f25ea7a1fb0db08011d2db/ui/app/containers/clusterOverview.tsx#L55)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/survive/tolerate/
> 
> this also seems like an odd place to put this warning - if any range has fewer than 3 nodes it should trigger a warning (this can happen via bad zone config, for instance). perhaps we could move this to a per-range warning?
</details>
This is different from a "range has fewer than 3 nodes" warning. This is "You are running fewer than 3 nodes." It's addressing https://github.com/cockroachdb/cockroach/issues/5978

---

*[ui/app/components/stackedgraph.tsx, line 79 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783#-KMV4-vvjSgLtEAVq5UZ:-KMVo4DuKEGMcYP8HB2H:-1921430770) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec759c23faa2d0199f25ea7a1fb0db08011d2db/ui/app/components/stackedgraph.tsx#L79)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> revert this.
</details>
Weird. I was getting an error. Do you know something about this?

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783#-:-KMVtFQD19E8eRxBQe1M:-1563279871)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 4 files at r1, 9 of 9 files at r4, 3 of 5 files at r8, 4 of 4 files at r9.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[ui/app/containers/clusterOverview.tsx, line 55 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783#-KMV4LYVUGhvMH3l1SGW:-KMVvAI0_-yAVDEX4Ct8:-1274347497) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec759c23faa2d0199f25ea7a1fb0db08011d2db/ui/app/containers/clusterOverview.tsx#L55)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> This is different from a "range has fewer than 3 nodes" warning. This is "You are running fewer than 3 nodes." It's addressing https://github.com/cockroachdb/cockroach/issues/5978
</details>
Yes, I understand that those are different, and yet they are the same. If a cluster has fewer than 3 nodes, all ranges necessarily have fewer than 3 replicas; I'm just wondering out loud if we should be describing the "real" concern that ranges are not sufficiently replicated.

BTW, we should tell the user that they need at least three, rather than saying that the current number they're using is not sufficient.

---

*[ui/app/components/stackedgraph.tsx, line 79 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783#-KMV4-vvjSgLtEAVq5UZ:-KMVv-d671Ux-LaAJgad:85526403) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec759c23faa2d0199f25ea7a1fb0db08011d2db/ui/app/components/stackedgraph.tsx#L79)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> Weird. I was getting an error. Do you know something about this?
> 
> Done.
</details>
Yes, these typings were recently updated.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[ui/app/components/sortabletable.tsx, line 125 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783#-KMVx-tc1rC5dibi2aXN:-KMVx-td6ssrgfHL1s9Z:461571833) ([raw file](https://github.com/cockroachdb/cockroach/blob/75fecf75a8825d2fb02230d016dafab5bca21374/ui/app/components/sortabletable.tsx#L125)):*
> ```XML
>           return <tr key={rowIndex}>
>             {_.map(columns, (c: SortableColumn, colIndex: number) => {
>               let content = c.cell(rowIndex);
> ```

Unless there is a very compelling reason, this is not appropriate logic for sortable table.  Can't this be a level up?  The providing `c.cell()` function already knows if it's a link or not, and should be able to add the `<div>` if necessary.

---

*[ui/app/components/visualization.tsx, line 9 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783#-KMVxp3zhjoxOWjwDP74:-KMVxp3zhjoxOWjwDP75:274695367) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad14b76707318549ef70d17ff447af6ea57aa814/ui/app/components/visualization.tsx#L9)):*
> ```XML
>   subtitle?: string;
>   tooltip?: string;
>   // If warning or warningTitle exist, they are appended to the tooltip
> ```

This isn't true, it requires `warning` to be set specifically.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783)*
<!-- Sent from Reviewable.io -->

maxlang commented:
Review status: 3 of 10 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[ui/app/components/sortabletable.tsx, line 125 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783#-KMVx-tc1rC5dibi2aXN:-KMVzZiUFL1CI8Xdaz-B:725308499) ([raw file](https://github.com/cockroachdb/cockroach/blob/75fecf75a8825d2fb02230d016dafab5bca21374/ui/app/components/sortabletable.tsx#L125)):*
<details><summary><i>Previously, mrtracy wrote…</i></summary>
> Unless there is a very compelling reason, this is not appropriate logic for sortable table.  Can't this be a level up?  The providing `c.cell()` function already knows if it's a link or not, and should be able to add the `<div>` if necessary.
</details>
It just feels like there's a greater chance for programmer error in that case. The `c.cell()` function is specified for each column. Why not centralize the logic?

---

*[ui/app/components/visualization.tsx, line 9 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783#-KMVxp3zhjoxOWjwDP74:-KM_HuwWtTaQ8HseTJTa:-1312588912) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad14b76707318549ef70d17ff447af6ea57aa814/ui/app/components/visualization.tsx#L9)):*
<details><summary><i>Previously, mrtracy wrote…</i></summary>
> This isn't true, it requires `warning` to be set specifically.
</details>
Fixed to make this true.

Done.

---

*[ui/app/containers/clusterOverview.tsx, line 55 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783#-KMV4LYVUGhvMH3l1SGW:-KMW-CFQGep9rQXUWIy-:722926972) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec759c23faa2d0199f25ea7a1fb0db08011d2db/ui/app/containers/clusterOverview.tsx#L55)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Yes, I understand that those are different, and yet they are the same. If a cluster has fewer than 3 nodes, all ranges necessarily have fewer than 3 replicas; I'm just wondering out loud if we should be describing the "real" concern that ranges are not sufficiently replicated.
> 
> BTW, we should tell the user that they need at least three, rather than saying that the current number they're using is not sufficient.
</details>
We should probably also 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783)*
<!-- Sent from Reviewable.io -->

maxlang commented:
Review status: 3 of 10 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[ui/app/containers/clusterOverview.tsx, line 55 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783#-KMV4LYVUGhvMH3l1SGW:-KM_JqOzqvh2t7IYHYyh:433550550) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec759c23faa2d0199f25ea7a1fb0db08011d2db/ui/app/containers/clusterOverview.tsx#L55)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> We should probably also 
</details>
Didn't finish my thought. We could probably do that as well, but since people often run 1 node clusters when they're just getting started, the idea was just to have some sort of indication that they're not getting the full benefit of Cockroach DB.

You're right that it should be based on the replication factor, which can change from range to range. We can make an issue to make this warning more explicit. For example: "You do not have enough nodes to support your highest replication factor (5). You must run at least 5 nodes."

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783)*
<!-- Sent from Reviewable.io -->

maxlang commented:
Review status: 3 of 10 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[ui/app/containers/clusterOverview.tsx, line 55 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783#-KMV4LYVUGhvMH3l1SGW:-KM_M02YKejcSc0fWXM3:-531343999) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec759c23faa2d0199f25ea7a1fb0db08011d2db/ui/app/containers/clusterOverview.tsx#L55)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> Didn't finish my thought. We could probably do that as well, but since people often run 1 node clusters when they're just getting started, the idea was just to have some sort of indication that they're not getting the full benefit of Cockroach DB.
> 
> You're right that it should be based on the replication factor, which can change from range to range. We can make an issue to make this warning more explicit. For example: "You do not have enough nodes to support your highest replication factor (5). You must run at least 5 nodes."
</details>
#7813

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
LGTM

---

Review status: 0 of 13 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783#-:-KM_xlwJKnnbTBY8lYAv:676834233)*
<!-- Sent from Reviewable.io -->

maxlang commented:
Review status: 0 of 13 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[ui/app/components/sortabletable.tsx, line 125 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783#-KMVx-tc1rC5dibi2aXN:-KM_xaUTKLFK_cac6X4F:-169528739) ([raw file](https://github.com/cockroachdb/cockroach/blob/75fecf75a8825d2fb02230d016dafab5bca21374/ui/app/components/sortabletable.tsx#L125)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> It just feels like there's a greater chance for programmer error in that case. The `c.cell()` function is specified for each column. Why not centralize the logic?
</details>
After discussing with @mrtracy, I decoupled the `Link` logic and added the ability to pass through a class name.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 13 of 13 files at r13, 5 of 5 files at r14, 4 of 4 files at r15.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[ui/app/containers/clusterOverview.tsx, line 55 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783#-KMV4LYVUGhvMH3l1SGW:-KMa-2CyRKQGuDvOYeo-:-1054777669) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec759c23faa2d0199f25ea7a1fb0db08011d2db/ui/app/containers/clusterOverview.tsx#L55)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> #7813
</details>
Again, we should at least (in this change) tell the user they need 3. Right now we are leaving them guessing.

---

*[ui/app/containers/nodesOverview.tsx, line 83 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783#-KM_zhIJ4q6_4jr0wS4Z:-KM_zhIKA6RIl5UmCE4U:1335552390) ([raw file](https://github.com/cockroachdb/cockroach/blob/45137df0d03bd6de53912cea7c843025648015ef/ui/app/containers/nodesOverview.tsx#L83)):*
> ```XML
>     key: NodesTableColumn.NodeID,
>     title: "Node",
>     cell: (ns) => <Link className="fill-cell" to={"/nodes/" + ns.desc.node_id}>{ns.desc.address.address_field}</Link>,
> ```

yikes, there is a lot of duplication in this first commit. no way around it?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[ui/app/containers/nodesOverview.tsx, line 83 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783#-KM_zhIJ4q6_4jr0wS4Z:-KMa-hEkiVA4x8MGdJG3:1193976377) ([raw file](https://github.com/cockroachdb/cockroach/blob/45137df0d03bd6de53912cea7c843025648015ef/ui/app/containers/nodesOverview.tsx#L83)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> yikes, there is a lot of duplication in this first commit. no way around it?
</details>
In my opinion, it would be overly cumbersome to de-duplicate this effectively. You could create a stateless component which maps to "Link" with className already set, but it looks like it's only being used in three places.  You would not be able to de-duplicate the `className` calls without considerable effort.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783)*
<!-- Sent from Reviewable.io -->

maxlang commented:
Review status: 2 of 13 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[ui/app/containers/clusterOverview.tsx, line 55 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783#-KMV4LYVUGhvMH3l1SGW:-KMdu3gBcPy18W_97oeF:-765958592) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec759c23faa2d0199f25ea7a1fb0db08011d2db/ui/app/containers/clusterOverview.tsx#L55)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Again, we should at least (in this change) tell the user they need 3. Right now we are leaving them guessing.
</details>
Changed the wording to:

```
A <number of nodes> node cluster cannot tolerate node failure. You need at least as many nodes as your largest replication factor, which is 3 by default.
```

cc @jseldess

---

*[ui/app/containers/nodesOverview.tsx, line 83 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783#-KM_zhIJ4q6_4jr0wS4Z:-KMa0AMHPQMWMw-uJC5R:196417570) ([raw file](https://github.com/cockroachdb/cockroach/blob/45137df0d03bd6de53912cea7c843025648015ef/ui/app/containers/nodesOverview.tsx#L83)):*
<details><summary><i>Previously, mrtracy wrote…</i></summary>
> In my opinion, it would be overly cumbersome to de-duplicate this effectively. You could create a stateless component which maps to "Link" with className already set, but it looks like it's only being used in three places.  You would not be able to de-duplicate the `className` calls without considerable effort.
</details>
Which duplication are you referring to? Some of it is because the various database pages already contain a lot of duplication, which should be refactored away at some point.

I changed from using a `no-padding` and `fill-cell` class to an `expand-link` on the `td` element, which will remove padding and expand a child `a` element.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783)*
<!-- Sent from Reviewable.io -->

maxlang commented:
Review status: 2 of 13 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[ui/app/containers/nodesOverview.tsx, line 83 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783#-KM_zhIJ4q6_4jr0wS4Z:-KMe5AL2q_D-Y1bSM_ig:925250841) ([raw file](https://github.com/cockroachdb/cockroach/blob/45137df0d03bd6de53912cea7c843025648015ef/ui/app/containers/nodesOverview.tsx#L83)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> Which duplication are you referring to? Some of it is because the various database pages already contain a lot of duplication, which should be refactored away at some point.
> 
> I changed from using a `no-padding` and `fill-cell` class to an `expand-link` on the `td` element, which will remove padding and expand a child `a` element.
</details>
It is true that `className: "expand-link"` is necessary every time there's a link, but it's nice to be explicit, rather than having hidden behavior just for the `Link` element as @mrtracy pointed out.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 4 of 11 files at r16, 3 of 5 files at r17, 4 of 4 files at r18.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[ui/app/containers/clusterOverview.tsx, line 55 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783#-KMV4LYVUGhvMH3l1SGW:-KMe5GWjtm1yaMQz8e4B:537740817) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec759c23faa2d0199f25ea7a1fb0db08011d2db/ui/app/containers/clusterOverview.tsx#L55)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> Changed the wording to:
> 
> ```
> A <number of nodes> node cluster cannot tolerate node failure. You need at least as many nodes as your largest replication factor, which is 3 by default.
> ```
> 
> cc @jseldess
</details>
No, this is also not correct. If your replication is set to 5, but you only have 3 nodes, you can still tolerate 1 machine failure.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783)*
<!-- Sent from Reviewable.io -->

maxlang commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[ui/app/containers/clusterOverview.tsx, line 55 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783#-KMV4LYVUGhvMH3l1SGW:-KMe8H0wmLCT9aWAAGlQ:1076982194) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec759c23faa2d0199f25ea7a1fb0db08011d2db/ui/app/containers/clusterOverview.tsx#L55)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> No, this is also not correct. If your replication is set to 5, but you only have 3 nodes, you can still tolerate 1 machine failure.
</details>
Really? Wouldn't you have quorum loss? I thought you needed at least `floor(R/2 +2)` nodes if you want to tolerate a node failure with replication factor R.

Regardless, the text is wrong. What about changing it to:

`A <number of nodes> node cluster cannot tolerate node failure. You need at least 3 nodes, and possibly more if you have increased the replication factor.`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[ui/app/containers/clusterOverview.tsx, line 55 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783#-KMV4LYVUGhvMH3l1SGW:-KMeC43Smd1k6tTk2mmo:450859116) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec759c23faa2d0199f25ea7a1fb0db08011d2db/ui/app/containers/clusterOverview.tsx#L55)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> Really? Wouldn't you have quorum loss? I thought you needed at least `floor(R/2 +2)` nodes if you want to tolerate a node failure with replication factor R.
> 
> Regardless, the text is wrong. What about changing it to:
> 
> `A <number of nodes> node cluster cannot tolerate node failure. You need at least 3 nodes, and possibly more if you have increased the replication factor.`
> 
</details>
```
Your cluster consists of <number> nodes, which is not sufficient to tolerate failure.

If you require fault tolerance, please increase the number of nodes in your cluster to at least 3.
```

The replication factor does not matter here (unless it is lower than 3).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783)*
<!-- Sent from Reviewable.io -->

jseldess commented:
*[ui/app/containers/clusterOverview.tsx, line 55 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783#-KMV4LYVUGhvMH3l1SGW:-KMeBWaF-umX_G-JfRps:-519417110) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec759c23faa2d0199f25ea7a1fb0db08011d2db/ui/app/containers/clusterOverview.tsx#L55)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> ```
> Your cluster consists of <number> nodes, which is not sufficient to tolerate failure.
> 
> If you require fault tolerance, please increase the number of nodes in your cluster to at least 3.
> ```
> 
> The replication factor does not matter here (unless it is lower than 3).
</details>
What about just: `You need at least 3 nodes to tolerate node failure. See Replica Recommendations for details.` where `Replica Recommendations` links to `https://www.cockroachlabs.com/docs/configure-replication-zones.html`...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[ui/app/containers/clusterOverview.tsx, line 55 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783#-KMV4LYVUGhvMH3l1SGW:-KMeEATTRrPov0ZPeWFg:341627940) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec759c23faa2d0199f25ea7a1fb0db08011d2db/ui/app/containers/clusterOverview.tsx#L55)):*
<details><summary><i>Previously, jseldess wrote…</i></summary>
> What about just: `You need at least 3 nodes to tolerate node failure. See Replica Recommendations for details.` where `Replica Recommendations` links to `https://www.cockroachlabs.com/docs/configure-replication-zones.html`...
</details>
Seems OK to me, as long as that document never moves.

However, I think we should avoid "you" - talk about "this cluster" rather than the operator.

I think that means:
```
This cluster consists of <number> nodes, which is not sufficient to tolerate failure.

If fault tolerance is required, please visit Replica Recommendations for best practices.
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783)*
<!-- Sent from Reviewable.io -->

maxlang commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[ui/app/containers/clusterOverview.tsx, line 55 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783#-KMV4LYVUGhvMH3l1SGW:-KMejUHyqKCVihRmmoch:-1575700438) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec759c23faa2d0199f25ea7a1fb0db08011d2db/ui/app/containers/clusterOverview.tsx#L55)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Seems OK to me, as long as that document never moves.
> 
> However, I think we should avoid "you" - talk about "this cluster" rather than the operator.
> 
> I think that means:
> ```
> This cluster consists of <number> nodes, which is not sufficient to tolerate failure.
> 
> If fault tolerance is required, please visit Replica Recommendations for best practices.
> ```
</details>
As a side note, the tooltips should be changed a bit to make clicking links on them easier.

I made the text:

```
<div>This cluster has fewer than 3 nodes, which is not sufficient to tolerate node failure.
              <br />See <a href="https://www.cockroachlabs.com/docs/configure-replication-zones.html" target="_blank">
                Replica Recommendations
              </a> in the docs for more details.</div>
```

Which is a combination of your approaches.

![Screen Shot 2016-07-14 at 2.55.59 PM.png](https://files.reviewableusercontent.io/images/reviewable/3691886/6vq2UlRKyK8GWYIDwSu_W4lx/Screen%20Shot%202016-07-14%20at%202.55.59%20PM.png)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783)*
<!-- Sent from Reviewable.io -->

jseldess commented:
*[ui/app/containers/clusterOverview.tsx, line 55 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783#-KMV4LYVUGhvMH3l1SGW:-KMemB-tYm-Ws9NZ9uGW:-258108531) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec759c23faa2d0199f25ea7a1fb0db08011d2db/ui/app/containers/clusterOverview.tsx#L55)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> As a side note, the tooltips should be changed a bit to make clicking links on them easier.
> 
> I made the text:
> 
> ```
> <div>This cluster has fewer than 3 nodes, which is not sufficient to tolerate node failure.
>               <br />See <a href="https://www.cockroachlabs.com/docs/configure-replication-zones.html" target="_blank">
>                 Replica Recommendations
>               </a> in the docs for more details.</div>
> ```
> 
> Which is a combination of your approaches.
> 
> ![Screen Shot 2016-07-14 at 2.55.59 PM.png](https://files.reviewableusercontent.io/images/reviewable/3691886/6vq2UlRKyK8GWYIDwSu_W4lx/Screen%20Shot%202016-07-14%20at%202.55.59%20PM.png)
> 
</details>
I'm ok with this, but I think we could still save space: "At least 3 nodes are required to tolerate node failure."

Also, on second thought, the link to the docs could be condensed: "See our docs for more details." where "docs" is the link. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783)*
<!-- Sent from Reviewable.io -->

maxlang commented:
Review status: 9 of 13 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[ui/app/containers/clusterOverview.tsx, line 55 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783#-KMV4LYVUGhvMH3l1SGW:-KMer5AJXAMKeqhsuUfM:11716285) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec759c23faa2d0199f25ea7a1fb0db08011d2db/ui/app/containers/clusterOverview.tsx#L55)):*
<details><summary><i>Previously, jseldess wrote…</i></summary>
> I'm ok with this, but I think we could still save space: "At least 3 nodes are required to tolerate node failure."
> 
> Also, on second thought, the link to the docs could be condensed: "See our docs for more details." where "docs" is the link. 
</details>
![Screen Shot 2016-07-14 at 3.22.30 PM.png](https://files.reviewableusercontent.io/images/reviewable/3691886/nH7RPi8dn3cbp9M0XJJlAMpG/Screen%20Shot%202016-07-14%20at%203.22.30%20PM.png)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 4 of 4 files at r19.
Review status: 11 of 13 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[ui/app/containers/clusterOverview.tsx, line 55 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783#-KMV4LYVUGhvMH3l1SGW:-KMerqMkMpJigaBQZ12a:-633712935) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec759c23faa2d0199f25ea7a1fb0db08011d2db/ui/app/containers/clusterOverview.tsx#L55)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> ![Screen Shot 2016-07-14 at 3.22.30 PM.png](https://files.reviewableusercontent.io/images/reviewable/3691886/nH7RPi8dn3cbp9M0XJJlAMpG/Screen%20Shot%202016-07-14%20at%203.22.30%20PM.png)
</details>
Am I missing something? Where is the number of nodes?

Also, is "At least 3 nodes are required to tolerate node failure" a sentence? Saving one line seems like a bad trade.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783)*
<!-- Sent from Reviewable.io -->

maxlang commented:
Review status: 11 of 13 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[ui/app/containers/clusterOverview.tsx, line 55 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783#-KMV4LYVUGhvMH3l1SGW:-KMesIno3MfNsYMArHfK:1910656989) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec759c23faa2d0199f25ea7a1fb0db08011d2db/ui/app/containers/clusterOverview.tsx#L55)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Am I missing something? Where is the number of nodes?
> 
> Also, is "At least 3 nodes are required to tolerate node failure" a sentence? Saving one line seems like a bad trade.
</details>
Yeah it's a full sentence.

The number of nodes is already visible in the giant number "visualization" to the left of this tooltip.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r20.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[ui/app/containers/clusterOverview.tsx, line 55 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783#-KMV4LYVUGhvMH3l1SGW:-KMeuJNwz2ZznCyYI9fO:1522494110) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec759c23faa2d0199f25ea7a1fb0db08011d2db/ui/app/containers/clusterOverview.tsx#L55)):*
<details><summary><i>Previously, maxlang (Max Lang) wrote…</i></summary>
> Yeah it's a full sentence.
> 
> The number of nodes is already visible in the giant number "visualization" to the left of this tooltip.
</details>
I see. That's not visible in the screenshot.

I don't think that's a sentence, but perhaps we've reached the limit of what can be discussed via comments.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783)*
<!-- Sent from Reviewable.io -->

jseldess commented:
*[ui/app/containers/clusterOverview.tsx, line 55 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783#-KMV4LYVUGhvMH3l1SGW:-KMf2XFIfsZhRwz7KDUM:-1396824617) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec759c23faa2d0199f25ea7a1fb0db08011d2db/ui/app/containers/clusterOverview.tsx#L55)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I see. That's not visible in the screenshot.
> 
> I don't think that's a sentence, but perhaps we've reached the limit of what can be discussed via comments.
</details>
Grammatically, it's definitely a sentence, but on further thought, I agree it's not as clear as the more verbose options. So let's go back to something closer to what Tamir suggested earlier:

`This cluster does not have enough nodes to tolerate failure.` 

And `If fault tolerance is required, see the docs for best practices.`with `docs` being the link.

Since the number of nodes is already displayed, we probably don't need to repeat it in the tooltip?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783)*
<!-- Sent from Reviewable.io -->

jseldess commented:
*[ui/app/containers/clusterOverview.tsx, line 55 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783#-KMV4LYVUGhvMH3l1SGW:-KMyTlaFMIeRqVFnCQAu:2121504699) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec759c23faa2d0199f25ea7a1fb0db08011d2db/ui/app/containers/clusterOverview.tsx#L55)):*
<details><summary><i>Previously, jseldess wrote…</i></summary>
> Grammatically, it's definitely a sentence, but on further thought, I agree it's not as clear as the more verbose options. So let's go back to something closer to what Tamir suggested earlier:
> 
> `This cluster does not have enough nodes to tolerate failure.` 
> 
> And `If fault tolerance is required, see the docs for best practices.`with `docs` being the link.
> 
> Since the number of nodes is already displayed, we probably don't need to repeat it in the tooltip?
</details>
OK, once more, this time mentioning the 3-node minimum:

`This cluster needs at least 3 nodes to tolerate failure. See the docs for more details.`

Hope it's ok with all? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783)*
<!-- Sent from Reviewable.io -->

maxlang commented:
Review status: 11 of 13 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[ui/app/containers/clusterOverview.tsx, line 55 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783#-KMV4LYVUGhvMH3l1SGW:-KMykfZe8l1jVImqyt7L:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/dec759c23faa2d0199f25ea7a1fb0db08011d2db/ui/app/containers/clusterOverview.tsx#L55)):*
<details><summary><i>Previously, jseldess wrote…</i></summary>
> OK, once more, this time mentioning the 3-node minimum:
> 
> `This cluster needs at least 3 nodes to tolerate failure. See the docs for more details.`
> 
> Hope it's ok with all? 
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7783)*
<!-- Sent from Reviewable.io -->
